engines/e_pkcs11_eng.c

```diff
@@ -0,0 +1,598 @@
+        if (pin != NULL)
+            ctx->pin = (CK_BYTE *) pin;
+    } else if (path == NULL) {
+       PKCS11_trace("inkey is null\n");
+       goto err;
+    } else {
+        if (ctx->module_path == NULL) {
+            PKCS11_trace("Module path is null\n");
+            goto err;
+        }
+        id = OPENSSL_strdup(path);
+
+        if (id == NULL) {
+            PKCS11err(PKCS11_F_PKCS11_PARSE, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        if (pkcs11_ishex(id))
+            id = pkcs11_hex2a(id);
+
+        slotid = NULL;
+    }
+
+    if (label != NULL)
+        ctx->label = (CK_BYTE *) label;
+    else
+        ctx->id = (CK_BYTE *) id;
+
+    if (slotid != NULL)
+        ctx->slotid = (CK_SLOT_ID) atoi(slotid);
+
+    if (ctx->pin == NULL) {
+        if (!pkcs11_get_console_pin(&pin))
+            goto err;
+        ctx->pin = (CK_BYTE *) pin;
+        if (ctx->pin == NULL) {
+            PKCS11_trace("PIN is invalid\n");
+            goto err;
+        }
+    }
+    return 1;
+
+ err:
+    return 0;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data)
+{
+    CK_RV rv;
+    PKCS11_CTX *ctx;
+
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+    if (!pkcs11_parse(ctx, path))
+        goto err;
+
+    rv = pkcs11_initialize(ctx->module_path);
+
+    if (rv != CKR_OK)
+        goto err;
+    if (!pkcs11_get_slot(ctx))
+        goto err;
+    if (!pkcs11_start_session(ctx))
+        goto err;
+    if (!pkcs11_login(ctx, CKU_USER))
+        goto err;
+    if (!pkcs11_find_private_key(ctx))
+        goto err;
+
+    return pkcs11_load_pkey(ctx);
+
+ err:
+    PKCS11_trace("pkcs11_engine_load_private_key failed\n");
+    return 0;
+}
+
+static OSSL_STORE_LOADER_CTX* pkcs11_store_open(
+    const OSSL_STORE_LOADER *loader, const char *uri,
+    const UI_METHOD *ui_method, void *ui_data)
+{
+    ENGINE *e;
+    PKCS11_CTX *pkcs11_ctx;
+    OSSL_STORE_LOADER_CTX *store_ctx = NULL;
+    char *objecttype = NULL, *object = NULL;
+
+    store_ctx = OSSL_STORE_LOADER_CTX_new();
+
+    e = (ENGINE *) OSSL_STORE_LOADER_get0_engine(loader);
+    if (e == NULL)
+        return NULL;
+
+    uri += 7;
+
+    pkcs11_parse_uri(uri,"objecttype=", &objecttype);
+    pkcs11_parse_uri(uri,"object=", &object);
+
+    if (object != NULL && objecttype != NULL
+        && strncmp(objecttype, "cert", 4) == 0) {
+        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+        if (pkcs11_ctx == NULL)
+            return NULL;
+        if (!pkcs11_get_cert(store_ctx, pkcs11_ctx, object))
+            return NULL;
+    }
+
+    if (strncmp(uri, "list=all", 8) == 0) {
+        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+        if (pkcs11_ctx == NULL)
+            return NULL;
+        if (!pkcs11_get_ids(store_ctx, pkcs11_ctx))
```

Comment:
So I looked into `pkcs11_get_ids`, and noticed the class `CKO_PUBLIC_KEY`, so ok, it lists any key and that's all.  But then above, you do not allow individual keys to be loaded...  That's kind weird...